compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -1385,20 +1410,17 @@ impl<'tcx> Predicate<'tcx> {
         }
     }
 
-    pub fn as_clause(self) -> Option<Binder<'tcx, ClauseKind<'tcx>>> {
-        let predicate = self.kind();
-        match predicate.skip_binder() {
-            PredicateKind::Clause(clause) => Some(predicate.rebind(clause)),
-            PredicateKind::AliasRelate(..)
-            | PredicateKind::Subtype(..)
-            | PredicateKind::Coerce(..)
-            | PredicateKind::ObjectSafe(..)
-            | PredicateKind::ClosureKind(..)
-            | PredicateKind::ConstEquate(..)
-            | PredicateKind::Ambiguous
-            | PredicateKind::TypeWellFormedFromEnv(..) => None,
+    pub fn as_clause(self) -> Option<Clause<'tcx>> {
+        match self.kind().skip_binder() {
+            PredicateKind::Clause(..) => Some(self.as_clause_unchecked()),
+            _ => None,
         }
     }
+
+    pub fn as_clause_unchecked(self) -> Clause<'tcx> {
+        debug_assert!(matches!(self.kind().skip_binder(), PredicateKind::Clause(_)));
```

Comment:
this assert is perf critical? and assuming that it is, would it not be faster to make this an assert and make using the clause use `unreachable_unchecked`?

Reply:
Anyways I made it an assert, and we can make it unreachable_unchecked on the read path if perf warrants it later.